src/BulkWriter/Pipelines/Pipeline.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using BulkWriter.Connect;
+using BulkWriter.Pipelines.Tasks;
+
+namespace BulkWriter.Pipelines
+{
+    public abstract class Pipeline
+    {
+        private readonly TaskFactory taskFactory = new TaskFactory(TaskCreationOptions.LongRunning, TaskContinuationOptions.None);
+        private readonly Stack<TaskBase> taskStack = new Stack<TaskBase>();
+        private readonly PipelineConfiguration pipelineConfiguration;
+
+        protected Pipeline(IConnectionProvider connectionProvider)
+        {
+            this.pipelineConfiguration = new PipelineConfiguration(connectionProvider, this.taskFactory, this.taskStack);
+        }
+
+        protected void SetUpPipeline(Action<IPipelineConfiguration> action)
+        {
+            action(this.pipelineConfiguration);
+        }
+
+        public virtual Task RunAsync(CancellationToken cancellationToken)
+        {
+            var writeAction = this.taskStack.Pop();
```

Comment:
It makes sense that a Write task needs to be the last one in the call chain, but I wonder if we ought to change the `IPipelineConfiguration` API to help enforce this?  At the very least, we can throw an exception here to ensure it's actually a `WriteTask`.